# WEEK THREE
### Etana Diasa <br> 11/15/2019
# Introduction
#### For this third week, we will download a dataset from the World Bank website. In this assignment we will do some data managing including handling NaN values and using the melt() function in order to reshape the data format, and also perform some descriptive analysis of the dataset. 

# Methods
## PART ONE: Importing and Exploring Data
#### To start working, the dataset we downloaded is 'Individual Using the Internet (% of population).

In [1]:
import pandas as pd
import numpy as np
MyData = pd.read_csv("WBData.csv")
MyData.shape

(222, 24)

#### This dataset has 222 rows and 24 columns. 

In [2]:
MyData.head()

Series Name     Series Code  \
0  Individuals using the Internet (% of population)  IT.NET.USER.ZS   
1  Individuals using the Internet (% of population)  IT.NET.USER.ZS   
2  Individuals using the Internet (% of population)  IT.NET.USER.ZS   
3  Individuals using the Internet (% of population)  IT.NET.USER.ZS   
4  Individuals using the Internet (% of population)  IT.NET.USER.ZS   

     Country Name Country Code 2000 [YR2000] 2001 [YR2001] 2002 [YR2002]  \
0     Afghanistan          AFG            ..   0.004722568   0.004561395   
1         Albania          ALB   0.114097347   0.325798377   0.390081273   
2         Algeria          DZA   0.491705679   0.646114017    1.59164126   
3  American Samoa          ASM            ..            ..            ..   
4         Andorra          AND   10.53883561            ..   11.26046872   

  2003 [YR2003] 2004 [YR2004] 2005 [YR2005]  ... 2010 [YR2010] 2011 [YR2011]  \
0   0.087891253    0.10580903   1.224148084  ...             4             5   
1   0.971900415   2.420387798   6.043890864  ...            45            49   
2   2.195359731   4.634475088   5.843942092  ...          12.5          14.9   
3            ..            ..            ..  ...            ..            ..   
4   13.54641288   26.83795439   37.60576622  ...            81            81   

  2012 [YR2012] 2013 [YR2013] 2014 [YR2014] 2015 [YR2015] 2016 [YR2016]  \
0   5.454545455           5.9             7          8.26          11.2   
1   54.65595904          57.2          60.1    63.2529327    66.3634447   
2          18.2          22.5          29.5          38.2   42.94552688   
3            ..            ..            ..            ..            ..   
4   86.43442462            94          95.9         96.91   97.93063712   

  2017 [YR2017] 2018 [YR2018] 2019 [YR2019]  
0          13.5            ..            ..  
1    71.8470405            ..            ..  
2   47.69105515   59.57968476            ..  
3            ..            ..            ..  
4   91.56746703            ..            ..  

[5 rows x 24 columns]

## Part Two: Managing Data
#### Among the 24 columns, the columns Series Name, Series Code and Country Code is a redundant information. We would remove those columns to create a working dataset. 

In [3]:
MyList = [2,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23]
MyData = MyData.iloc[:, MyList]
MyData.tail(10)

Country Name 2000 [YR2000] 2001 [YR2001] 2002 [YR2002]  \
212  Virgin Islands (U.S.)   13.81508054   18.37576604   27.49443238   
213     West Bank and Gaza   1.111306207   1.836854839   3.100092235   
214            Yemen, Rep.   0.082500387   0.090802459   0.518796032   
215                 Zambia   0.191071643   0.233129556   0.477750907   
216               Zimbabwe   0.401433535   0.799846046           1.1   
217                    NaN           NaN           NaN           NaN   
218                    NaN           NaN           NaN           NaN   
219                    NaN           NaN           NaN           NaN   
220                    NaN           NaN           NaN           NaN   
221                    NaN           NaN           NaN           NaN   

    2003 [YR2003] 2004 [YR2004] 2005 [YR2005] 2006 [YR2006] 2007 [YR2007]  \
212   27.42907299   27.37700879   27.34431967   27.33261054   27.33933584   
213   4.130616356   4.400904826        16.005         18.41        21.176   
214     0.6047341   0.881222989   1.048597934   1.247824049          5.01   
215   0.980483039   2.013549532   2.851752261   4.159913394          4.87   
216           1.8           2.1           2.4           2.4             3   
217           NaN           NaN           NaN           NaN           NaN   
218           NaN           NaN           NaN           NaN           NaN   
219           NaN           NaN           NaN           NaN           NaN   
220           NaN           NaN           NaN           NaN           NaN   
221           NaN           NaN           NaN           NaN           NaN   

    2008 [YR2008]  ... 2010 [YR2010] 2011 [YR2011] 2012 [YR2012]  \
212   27.36177742  ...         31.22          35.6       40.5479   
213        24.358  ...          37.4         41.08          43.4   
214          6.89  ...         12.35        14.905       17.4465   
215          5.55  ...            10          11.5       13.4682   
216           3.5  ...           6.4           8.4            12   
217           NaN  ...           NaN           NaN           NaN   
218           NaN  ...           NaN           NaN           NaN   
219           NaN  ...           NaN           NaN           NaN   
220           NaN  ...           NaN           NaN           NaN   
221           NaN  ...           NaN           NaN           NaN   

    2013 [YR2013] 2014 [YR2014] 2015 [YR2015] 2016 [YR2016] 2017 [YR2017]  \
212          45.3         50.07    54.8391373    59.6083158    64.3774943   
213          46.6         53.67   57.42419241   61.17838483          65.2   
214            20         22.55   24.08540947   24.57920836   26.71835477   
215          15.4            19            21   25.50657885    27.8525791   
216          15.5   16.36473996    22.7428181   23.11998904   27.05548769   
217           NaN           NaN           NaN           NaN           NaN   
218           NaN           NaN           NaN           NaN           NaN   
219           NaN           NaN           NaN           NaN           NaN   
220           NaN           NaN           NaN           NaN           NaN   
221           NaN           NaN           NaN           NaN           NaN   

    2018 [YR2018] 2019 [YR2019]  
212            ..            ..  
213   64.39999371            ..  
214            ..            ..  
215    14.2999971            ..  
216            ..            ..  
217           NaN           NaN  
218           NaN           NaN  
219           NaN           NaN  
220           NaN           NaN  
221           NaN           NaN  

[10 rows x 21 columns]

#### There are several rows that are empty and of no particular value for the use of this dataset. Therefore, I will remove (row 216 and below) those entries in this next line of code. 

In [4]:
MyData = MyData.loc[0:216, :]
MyData.tail()

Country Name 2000 [YR2000] 2001 [YR2001] 2002 [YR2002]  \
212  Virgin Islands (U.S.)   13.81508054   18.37576604   27.49443238   
213     West Bank and Gaza   1.111306207   1.836854839   3.100092235   
214            Yemen, Rep.   0.082500387   0.090802459   0.518796032   
215                 Zambia   0.191071643   0.233129556   0.477750907   
216               Zimbabwe   0.401433535   0.799846046           1.1   

    2003 [YR2003] 2004 [YR2004] 2005 [YR2005] 2006 [YR2006] 2007 [YR2007]  \
212   27.42907299   27.37700879   27.34431967   27.33261054   27.33933584   
213   4.130616356   4.400904826        16.005         18.41        21.176   
214     0.6047341   0.881222989   1.048597934   1.247824049          5.01   
215   0.980483039   2.013549532   2.851752261   4.159913394          4.87   
216           1.8           2.1           2.4           2.4             3   

    2008 [YR2008]  ... 2010 [YR2010] 2011 [YR2011] 2012 [YR2012]  \
212   27.36177742  ...         31.22          35.6       40.5479   
213        24.358  ...          37.4         41.08          43.4   
214          6.89  ...         12.35        14.905       17.4465   
215          5.55  ...            10          11.5       13.4682   
216           3.5  ...           6.4           8.4            12   

    2013 [YR2013] 2014 [YR2014] 2015 [YR2015] 2016 [YR2016] 2017 [YR2017]  \
212          45.3         50.07    54.8391373    59.6083158    64.3774943   
213          46.6         53.67   57.42419241   61.17838483          65.2   
214            20         22.55   24.08540947   24.57920836   26.71835477   
215          15.4            19            21   25.50657885    27.8525791   
216          15.5   16.36473996    22.7428181   23.11998904   27.05548769   

    2018 [YR2018] 2019 [YR2019]  
212            ..            ..  
213   64.39999371            ..  
214            ..            ..  
215    14.2999971            ..  
216            ..            ..  

[5 rows x 21 columns]

## Part 3: Sorting Through NaN Values
#### There are several NaN values that were entered as "..". It would be easy to convert those values to NaN now and handle them later after we melt the dataset. 

In [5]:
MyData = MyData.replace(".." , np.nan)
MyData

Country Name 2000 [YR2000] 2001 [YR2001] 2002 [YR2002]  \
0              Afghanistan           NaN   0.004722568   0.004561395   
1                  Albania   0.114097347   0.325798377   0.390081273   
2                  Algeria   0.491705679   0.646114017    1.59164126   
3           American Samoa           NaN           NaN           NaN   
4                  Andorra   10.53883561           NaN   11.26046872   
..                     ...           ...           ...           ...   
212  Virgin Islands (U.S.)   13.81508054   18.37576604   27.49443238   
213     West Bank and Gaza   1.111306207   1.836854839   3.100092235   
214            Yemen, Rep.   0.082500387   0.090802459   0.518796032   
215                 Zambia   0.191071643   0.233129556   0.477750907   
216               Zimbabwe   0.401433535   0.799846046           1.1   

    2003 [YR2003] 2004 [YR2004] 2005 [YR2005] 2006 [YR2006] 2007 [YR2007]  \
0     0.087891253    0.10580903   1.224148084   2.107123645           1.9   
1     0.971900415   2.420387798   6.043890864   9.609991316   15.03611541   
2     2.195359731   4.634475088   5.843942092   7.375984956   9.451190626   
3             NaN           NaN           NaN           NaN           NaN   
4     13.54641288   26.83795439   37.60576622     48.936847         70.87   
..            ...           ...           ...           ...           ...   
212   27.42907299   27.37700879   27.34431967   27.33261054   27.33933584   
213   4.130616356   4.400904826        16.005         18.41        21.176   
214     0.6047341   0.881222989   1.048597934   1.247824049          5.01   
215   0.980483039   2.013549532   2.851752261   4.159913394          4.87   
216           1.8           2.1           2.4           2.4             3   

    2008 [YR2008]  ... 2010 [YR2010] 2011 [YR2011] 2012 [YR2012]  \
0            1.84  ...             4             5   5.454545455   
1           23.86  ...            45            49   54.65595904   
2           10.18  ...          12.5          14.9          18.2   
3             NaN  ...           NaN           NaN           NaN   
4           70.04  ...            81            81   86.43442462   
..            ...  ...           ...           ...           ...   
212   27.36177742  ...         31.22          35.6       40.5479   
213        24.358  ...          37.4         41.08          43.4   
214          6.89  ...         12.35        14.905       17.4465   
215          5.55  ...            10          11.5       13.4682   
216           3.5  ...           6.4           8.4            12   

    2013 [YR2013] 2014 [YR2014] 2015 [YR2015] 2016 [YR2016] 2017 [YR2017]  \
0             5.9             7          8.26          11.2          13.5   
1            57.2          60.1    63.2529327    66.3634447    71.8470405   
2            22.5          29.5          38.2   42.94552688   47.69105515   
3             NaN           NaN           NaN           NaN           NaN   
4              94          95.9         96.91   97.93063712   91.56746703   
..            ...           ...           ...           ...           ...   
212          45.3         50.07    54.8391373    59.6083158    64.3774943   
213          46.6         53.67   57.42419241   61.17838483          65.2   
214            20         22.55   24.08540947   24.57920836   26.71835477   
215          15.4            19            21   25.50657885    27.8525791   
216          15.5   16.36473996    22.7428181   23.11998904   27.05548769   

    2018 [YR2018] 2019 [YR2019]  
0             NaN           NaN  
1             NaN           NaN  
2     59.57968476           NaN  
3             NaN           NaN  
4             NaN           NaN  
..            ...           ...  
212           NaN           NaN  
213   64.39999371           NaN  
214           NaN           NaN  
215    14.2999971           NaN  
216           NaN           NaN  

[217 rows x 21 columns]

#### Now that we have converted every missing value as NaN, we can go ahead and transform our table. I chose to deal with missing values for two reasons. 1) I did not want to use the median() method because the table puts consequitive internet usage data across the table as oppose to top to bottom. Therefore, rather than the mean population across the years, it would put the mean value of all countries per year. 2) The Forward and Backward filling would also capture the internet usage data of the neighbor country listed before or after a particular country, as oppose to a proceeding or preceeding year of the same country which is what we desire to doM. I hope this paragraph makes sense. :) Now, let us melt the datatable with 'Country Name' being our id_variable and all the remaining columns columns[1:] being our vairables, we will not populate the % of Population as our values. Below is what the table looks like. 

In [11]:
MyMeltedData = pd.melt(MyData, id_vars= 'Country Name', value_vars = list(MyData.columns[1:]), var_name="Year", value_name="% of Population")
MyMeltedData

Country Name           Year % of Population
0               Afghanistan  2000 [YR2000]             NaN
1                   Albania  2000 [YR2000]     0.114097347
2                   Algeria  2000 [YR2000]     0.491705679
3            American Samoa  2000 [YR2000]             NaN
4                   Andorra  2000 [YR2000]     10.53883561
...                     ...            ...             ...
4335  Virgin Islands (U.S.)  2019 [YR2019]             NaN
4336     West Bank and Gaza  2019 [YR2019]             NaN
4337            Yemen, Rep.  2019 [YR2019]             NaN
4338                 Zambia  2019 [YR2019]             NaN
4339               Zimbabwe  2019 [YR2019]             NaN

[4340 rows x 3 columns]

#### Our new table Melted Data has 4340 rows and 3 columns. It pupulates each years of % of population data for every country. Now, in order to deal with the missing values, we would need to make the countries as our columns and the year 2000 to 2019 as our index for the year. I use the pivot() function to do exactly that. 

In [7]:
MyPivotedData = MeltedData.pivot(index='Year', columns = 'Country Name', values="% of Population")
MyPivotedData

Country Name   Afghanistan      Albania      Algeria American Samoa  \
Year                                                                  
2000 [YR2000]          NaN  0.114097347  0.491705679            NaN   
2001 [YR2001]  0.004722568  0.325798377  0.646114017            NaN   
2002 [YR2002]  0.004561395  0.390081273   1.59164126            NaN   
2003 [YR2003]  0.087891253  0.971900415  2.195359731            NaN   
2004 [YR2004]   0.10580903  2.420387798  4.634475088            NaN   
2005 [YR2005]  1.224148084  6.043890864  5.843942092            NaN   
2006 [YR2006]  2.107123645  9.609991316  7.375984956            NaN   
2007 [YR2007]          1.9  15.03611541  9.451190626            NaN   
2008 [YR2008]         1.84        23.86        10.18            NaN   
2009 [YR2009]         3.55         41.2        11.23            NaN   
2010 [YR2010]            4           45         12.5            NaN   
2011 [YR2011]            5           49         14.9            NaN   
2012 [YR2012]  5.454545455  54.65595904         18.2            NaN   
2013 [YR2013]          5.9         57.2         22.5            NaN   
2014 [YR2014]            7         60.1         29.5            NaN   
2015 [YR2015]         8.26   63.2529327         38.2            NaN   
2016 [YR2016]         11.2   66.3634447  42.94552688            NaN   
2017 [YR2017]         13.5   71.8470405  47.69105515            NaN   
2018 [YR2018]          NaN          NaN  59.57968476            NaN   
2019 [YR2019]          NaN          NaN          NaN            NaN   

Country Name       Andorra       Angola Antigua and Barbuda    Argentina  \
Year                                                                       
2000 [YR2000]  10.53883561  0.105045562         6.482225737  7.038683086   
2001 [YR2001]          NaN  0.136013867         8.899285515  9.780807285   
2002 [YR2002]  11.26046872  0.270376746                12.5  10.88212438   
2003 [YR2003]  13.54641288  0.370682065         17.22864878  11.91369655   
2004 [YR2004]  26.83795439  0.464814618         24.26654372  16.03668411   
2005 [YR2005]  37.60576622  1.143366827                  27  17.72058337   
2006 [YR2006]    48.936847          1.5                  30   20.9272021   
2007 [YR2007]        70.87          1.7                  34  25.94663294   
2008 [YR2008]        70.04          1.9                  38  28.11262348   
2009 [YR2009]        78.53          2.3                  42           34   
2010 [YR2010]           81          2.8                  47           45   
2011 [YR2011]           81          3.1                  52           51   
2012 [YR2012]  86.43442462          6.5                  58         55.8   
2013 [YR2013]           94          8.9                63.4         59.9   
2014 [YR2014]         95.9         21.4               67.78         64.7   
2015 [YR2015]        96.91         12.4                  70  68.04306411   
2016 [YR2016]  97.93063712           13                  73  70.96898082   
2017 [YR2017]  91.56746703  14.33907936                  76  74.29490687   
2018 [YR2018]          NaN          NaN                 NaN          NaN   
2019 [YR2019]          NaN          NaN                 NaN          NaN   

Country Name       Armenia        Aruba  ...      Uruguay   Uzbekistan  \
Year                                     ...                             
2000 [YR2000]  1.300470022  15.44282295  ...  10.53905775  0.484347308   
2001 [YR2001]  1.631094667         17.1  ...  11.12143768  0.597567994   
2002 [YR2002]  1.960405046         18.8  ...   11.4194702  1.081939742   
2003 [YR2003]  4.575217225         20.8  ...  15.93713672  1.912595178   
2004 [YR2004]  4.899008571           23  ...  17.06309834  2.593725421   
2005 [YR2005]  5.252983352         25.4  ...  20.08818956  3.343509888   
2006 [YR2006]  5.631787777           28  ...         29.4  6.388321967   
2007 [YR2007]  6.021253397         30.9  ...           34  7.490604681   
2008 [YR2008]         6.

#### Our table now looks ready to have a bfill() and ffill() to take care of missing elements. If we have a column that remains NaN, means there was not a single entry for that country. Let us find out what our new table looks like. We call it MyCleanData

In [8]:
MyCleanData = MyPivotedData.fillna(method='bfill')
MyCleanData = MyCleanData.fillna(method='ffill')
MyCleanData

Country Name   Afghanistan      Albania      Algeria  American Samoa  \
Year                                                                   
2000 [YR2000]  0.004722568  0.114097347  0.491705679             NaN   
2001 [YR2001]  0.004722568  0.325798377  0.646114017             NaN   
2002 [YR2002]  0.004561395  0.390081273   1.59164126             NaN   
2003 [YR2003]  0.087891253  0.971900415  2.195359731             NaN   
2004 [YR2004]   0.10580903  2.420387798  4.634475088             NaN   
2005 [YR2005]  1.224148084  6.043890864  5.843942092             NaN   
2006 [YR2006]  2.107123645  9.609991316  7.375984956             NaN   
2007 [YR2007]          1.9  15.03611541  9.451190626             NaN   
2008 [YR2008]         1.84        23.86        10.18             NaN   
2009 [YR2009]         3.55         41.2        11.23             NaN   
2010 [YR2010]            4           45         12.5             NaN   
2011 [YR2011]            5           49         14.9             NaN   
2012 [YR2012]  5.454545455  54.65595904         18.2             NaN   
2013 [YR2013]          5.9         57.2         22.5             NaN   
2014 [YR2014]            7         60.1         29.5             NaN   
2015 [YR2015]         8.26   63.2529327         38.2             NaN   
2016 [YR2016]         11.2   66.3634447  42.94552688             NaN   
2017 [YR2017]         13.5   71.8470405  47.69105515             NaN   
2018 [YR2018]         13.5   71.8470405  59.57968476             NaN   
2019 [YR2019]         13.5   71.8470405  59.57968476             NaN   

Country Name       Andorra       Angola Antigua and Barbuda    Argentina  \
Year                                                                       
2000 [YR2000]  10.53883561  0.105045562         6.482225737  7.038683086   
2001 [YR2001]  11.26046872  0.136013867         8.899285515  9.780807285   
2002 [YR2002]  11.26046872  0.270376746                12.5  10.88212438   
2003 [YR2003]  13.54641288  0.370682065         17.22864878  11.91369655   
2004 [YR2004]  26.83795439  0.464814618         24.26654372  16.03668411   
2005 [YR2005]  37.60576622  1.143366827                  27  17.72058337   
2006 [YR2006]    48.936847          1.5                  30   20.9272021   
2007 [YR2007]        70.87          1.7                  34  25.94663294   
2008 [YR2008]        70.04          1.9                  38  28.11262348   
2009 [YR2009]        78.53          2.3                  42           34   
2010 [YR2010]           81          2.8                  47           45   
2011 [YR2011]           81          3.1                  52           51   
2012 [YR2012]  86.43442462          6.5                  58         55.8   
2013 [YR2013]           94          8.9                63.4         59.9   
2014 [YR2014]         95.9         21.4               67.78         64.7   
2015 [YR2015]        96.91         12.4                  70  68.04306411   
2016 [YR2016]  97.93063712           13                  73  70.96898082   
2017 [YR2017]  91.56746703  14.33907936                  76  74.29490687   
2018 [YR2018]  91.56746703  14.33907936                  76  74.29490687   
2019 [YR2019]  91.56746703  14.33907936                  76  74.29490687   

Country Name       Armenia        Aruba  ...      Uruguay   Uzbekistan  \
Year                                     ...                             
2000 [YR2000]  1.300470022  15.44282295  ...  10.53905775  0.484347308   
2001 [YR2001]  1.631094667         17.1  ...  11.12143768  0.597567994   
2002 [YR2002]  1.960405046         18.8  ...   11.4194702  1.081939742   
2003 [YR2003]  4.575217225         20.8  ...  15.93713672  1.912595178   
2004 [YR2004]  4.899008571           23  ...  17.06309834  2.593725421   
2005 [YR2005]  5.252983352         25.4  ...  20.08818956  3.343509888   
2006 [YR2006]  5.631787777           28  ...         29.4  6.388321967   
2007 [YR2007]  6.021253397         30.9  ...           34  7.490604681   
20

#### Before we go to the process of analyzing the data, I had also come across a much simpler way of transposing tables. 

In [9]:
MyTransposedData = MyData.T
MyTransposedData.head()

0            1            2               3    \
Country Name   Afghanistan      Albania      Algeria  American Samoa   
2000 [YR2000]          NaN  0.114097347  0.491705679             NaN   
2001 [YR2001]  0.004722568  0.325798377  0.646114017             NaN   
2002 [YR2002]  0.004561395  0.390081273   1.59164126             NaN   
2003 [YR2003]  0.087891253  0.971900415  2.195359731             NaN   

                       4            5                    6            7    \
Country Name       Andorra       Angola  Antigua and Barbuda    Argentina   
2000 [YR2000]  10.53883561  0.105045562          6.482225737  7.038683086   
2001 [YR2001]          NaN  0.136013867          8.899285515  9.780807285   
2002 [YR2002]  11.26046872  0.270376746                 12.5  10.88212438   
2003 [YR2003]  13.54641288  0.370682065          17.22864878  11.91369655   

                       8            9    ...          207          208  \
Country Name       Armenia        Aruba  ...      Uruguay   Uzbekistan   
2000 [YR2000]  1.300470022  15.44282295  ...  10.53905775  0.484347308   
2001 [YR2001]  1.631094667         17.1  ...  11.12143768  0.597567994   
2002 [YR2002]  1.960405046         18.8  ...   11.4194702  1.081939742   
2003 [YR2003]  4.575217225         20.8  ...  15.93713672  1.912595178   

                       209            210          211                    212  \
Country Name       Vanuatu  Venezuela, RB      Vietnam  Virgin Islands (U.S.)   
2000 [YR2000]  2.108336891    3.359597465  0.254248276            13.81508054   
2001 [YR2001]  2.830572239    4.636000948  1.265651236            18.37576604   
2002 [YR2002]  3.510038711    4.910446327  1.854999236            27.49443238   
2003 [YR2003]  3.903295845    7.499963465  3.780280814            27.42907299   

                              213          214          215          216  
Country Name   West Bank and Gaza  Yemen, Rep.       Zambia     Zimbabwe  
2000 [YR2000]         1.111306207  0.082500387  0.191071643  0.401433535  
2001 [YR2001]         1.836854839  0.090802459  0.233129556  0.799846046  
2002 [YR2002]         3.100092235  0.518796032  0.477750907          1.1  
2003 [YR2003]         4.130616356    0.6047341  0.980483039          1.8  

[5 rows x 217 columns]

## Part 4: Simple Data Analysis
#### Now, using the clean data I have, I would melt it to do some analysis. 

In [54]:
MyCleanMelted = pd.melt(MyCleanData, value_vars = list(MyCleanData.columns[1:]), var_name="Country", value_name="Population")
MyCleanMelted

Country   Population
0      Albania  0.114097347
1      Albania  0.325798377
2      Albania  0.390081273
3      Albania  0.971900415
4      Albania  2.420387798
...        ...          ...
4315  Zimbabwe   22.7428181
4316  Zimbabwe  23.11998904
4317  Zimbabwe  27.05548769
4318  Zimbabwe  27.05548769
4319  Zimbabwe  27.05548769

[4320 rows x 2 columns]

#### First, I will change the datatype of the 'Population' column into numeric in order to perform basic numerical analysis. 

In [89]:
MyCleanMelted["Population"] = pd.to_numeric(MyCleanMelted["Population"])
MyCleanMelted.dtypes

Country        object
Population    float64
dtype: object

#### Below, using the groupby() function, I have attempted to find the median % of population that have access to the internet per each country there is data. Here, American Samoa does not have data. 

In [101]:
MyMax = MyCleanMelted.groupby(['Country']) \
[['Population']]. \
max()
MyMax = MyMax.rename(columns={"Population" : "Max"})
MyMax.head(5)

Max
Country                  
Albania         71.847041
Algeria         59.579685
American Samoa        NaN
Andorra         97.930637
Angola          21.400000

In [102]:
MyMin = MyCleanMelted.groupby(['Country']) \
[['Population']]. \
min()
MyMin = MyMin.rename(columns={"Population" : "Min"})
MyMin.head(5)

Min
Country                  
Albania          0.114097
Algeria          0.491706
American Samoa        NaN
Andorra         10.538836
Angola           0.105046

In [115]:
MyNewTable = pd.merge(MyMin,MyMax, on='Country', how='left')
MyNewTable["Growth"] = MyNewTable["Max"] - MyNewTable["Min"]
MyNewTable

Min        Max     Growth
Country                                               
Albania                 0.114097  71.847041  71.732943
Algeria                 0.491706  59.579685  59.087979
American Samoa               NaN        NaN        NaN
Andorra                10.538836  97.930637  87.391802
Angola                  0.105046  21.400000  21.294954
...                          ...        ...        ...
Virgin Islands (U.S.)  13.815081  64.377494  50.562414
West Bank and Gaza      1.111306  65.200000  64.088694
Yemen, Rep.             0.082500  26.718355  26.635854
Zambia                  0.191072  27.852579  27.661507
Zimbabwe                0.401434  27.055488  26.654054

[216 rows x 3 columns]

#### Using simple analysis technique, we can identify which country has the largest growth in population of internet users and vice versa. 

In [116]:
MyNewTable[MyNewTable.Growth == MyNewTable.Growth.max()]

Min        Max    Growth
Country                               
Qatar    4.863679  99.652849  94.78917

#### It appears that Qatar had a tremendous growth in growing its internet user population significanly. 

In [117]:
MyNewTable[MyNewTable.Growth == MyNewTable.Growth.min()]

Min  Max  Growth
Country                                    
Korea, Dem. People’s Rep.  0.0  0.0     0.0

#### Per our finding, it appears that Dem. People's Rep. of Korea has the slowest growth. However, it is the case that a value of 0.0 has been recorded for the 20 years of datagathering. Therefore, to be fair, I will delete this row and recalculate my findings one more time.  

In [128]:
MyNewTable = MyNewTable[MyNewTable.Max != 0.0]
#MyMedian[MyMedian.Growth == MyMedian.Growth.min()]
MyNewTable

Min        Max     Growth
Country                                               
Albania                 0.114097  71.847041  71.732943
Algeria                 0.491706  59.579685  59.087979
American Samoa               NaN        NaN        NaN
Andorra                10.538836  97.930637  87.391802
Angola                  0.105046  21.400000  21.294954
...                          ...        ...        ...
Virgin Islands (U.S.)  13.815081  64.377494  50.562414
West Bank and Gaza      1.111306  65.200000  64.088694
Yemen, Rep.             0.082500  26.718355  26.635854
Zambia                  0.191072  27.852579  27.661507
Zimbabwe                0.401434  27.055488  26.654054

[215 rows x 3 columns]

In [129]:
MyNewTable[MyNewTable.Growth == MyNewTable.Growth.min()]

Min       Max    Growth
Country                              
Eritrea  0.136712  1.308907  1.172195

#### After removing 0.0 values from Max column, it appears that Eritrea is the country with the lowest growth grate of almost 1 percent increase in her population of internet users in 20 years span. 

# Conclusion
#### Pandas and NamPy provide useful tools to manage data and provide simply analysis on a given dataset. For this assignment, we were able to:
#### 1. Import dataset and explore the dataset. 
#### 2. Managing Data
#### 3. Sorthing through NaN Value
#### 4. Simple Data Analysis

#### According to this assignment, data management can be performed in various ways. The melt() and pivot() functions provided what Transposing function is able to do. Furthermore, ffill() and bfill() functions provide a very useful medium in order to deal with missing or NaN values. However, the challenge was applying these functions horizontally than vertically. 

#### The last challenge was being able to sort through the different table I had created and managed to merge them. Also, taking out "0.0" data entries was another assignment for this project. 

# References

#### https://stackoverflow.com/questions/18172851/deleting-dataframe-row-in-pandas-based-on-column-value
#### https://stackoverflow.com/questions/27764378/how-to-reverse-a-2-dimensional-table-dataframe-into-a-1-dimensional-list-using
#### https://chrisalbon.com/python/data_wrangling/pandas_replace_values/